src/System.Management.Automation/engine/remoting/fanin/OutOfProcTransportManager.cs

```diff
@@ -1641,6 +1650,43 @@ internal override void CreateAsync()
 
             // Create reader thread and send first PSRP message.
             StartReaderThread(_stdOutReader);
+
+            // Monitor SSH process until connection is established.
+            _sshProcessTimer = new Timer(
+                callback: (_) =>
+                {
+                    if (_connectionEstablished)
+                    {
+                        // Monitor no longer needed.
+                        _sshProcessTimer?.Change(Timeout.Infinite, Timeout.Infinite);
+                        return;
+                    }
+
+                    bool sshTerminated = false;
+                    try
+                    {
+                        // Detect if SSH client process terminates prematurely.
+                        using (var sshProcess = System.Diagnostics.Process.GetProcessById(_sshProcessId))
+                        {
+                            sshTerminated = sshProcess == null || sshProcess.Handle == IntPtr.Zero || sshProcess.HasExited;
+                        }
+                    }
+                    catch
+                    {
+                        sshTerminated = true;
+                    }
+
+                    if (sshTerminated)
+                    {
+                        _sshProcessTimer?.Change(Timeout.Infinite, Timeout.Infinite);
+                        _sshProcessId = 0;
+                        HandleSSHError(
+                            new PSRemotingTransportException(RemotingErrorIdStrings.SSHClientCannotConnect));
+                    }
+                },
+                state: null,
+                dueTime: 100,   // 100 mSec
+                period: 100);   // Repeat indefinitely
```

Comment:
waiting forever always makes me nervous. Should we have some sort of timeout here?
If I read this correctly, once we have a connection, we stop watching, but if the connection is never established, I suppose we would never stop checking.

Reply:
Having a timeout for connect is always problematic because it is impossible to know what a reasonable value is.  There is always some situation that violates it.  So we never do it for a connection, however we do have a close connection timeout, because it can be handled relatively safely.  This fix is just for a known condition that we want to catch.

Note that the connection attempt can always be cancelled with `Ctrl+C`.  But this doesn't help in non-interactive automation cases.

Reply:
This raises concerns for me too.
There are two thoughts:
1. A suitable timeout might be TCP/IP timeout ~30 sec.
2. If we are going to move Remoting to a subsystem it would be better to have a more general solution suitable for _any_ transport. Is it possible to do something at a higher level? How does PSRP determine if a connection is up or down? Can we add a hung check there?

Reply:
I definitely won't put in a hard coded timeout time.  But we could add a new SSH remoting parameter that can optionally set a maximum time to wait for a connection, and which value defaults to infinity.  This feels like the best solution.  As I mention above an SSH connection attempt can be cancelled at any time with `Ctrl+C`.  The new parameter would be for unattended script.

The remoting protocol does not deal with initial connection and assumes a connection is established when it begins.  Establishing and authenticating a connection is left to each individual implementation.

I'll remove the timer in this PR and just leave the message queue clean up fix.
I'll create a new PR for the SSH connection attempt timeout feature.

Reply:
> The new parameter would be for unattended script.

Original issue comes from CI fail due to timeout. This implies that we will have to always use the new parameter and set this timeout. Then why not put some default reasonable value for _convenience_?

Reply:
Because everyone has a different idea of what 'reasonable' is.  We have debated this many times on the team and have concluded to never impose a hard coded limitation unless unavoidable.

Instead of breaking this into two PRs, I think I'll just keep the changes to this single one.
I'll tag remoting working group to review the proposal.

Reply:
> We have debated this many times on the team

Accepted! Thanks!

Will we have to use the new parameter everywhere in tests to avoid CI hangs?

Reply:
Yes, SSH tests will be updated.